iree/hal/cuda/cuda_driver.c

```diff
@@ -161,22 +180,23 @@ static iree_status_t iree_hal_cuda_driver_query_available_devices(
 }
 
 static iree_status_t iree_hal_cuda_driver_select_default_device(
-    iree_hal_cuda_dynamic_symbols_t* syms, int default_device_index,
-    iree_allocator_t host_allocator, CUdevice* out_device) {
-  int device_count = 0;
-  CUDA_RETURN_IF_ERROR(syms, cuDeviceGetCount(&device_count),
-                       "cuDeviceGetCount");
+    iree_hal_driver_t* base_driver, iree_hal_cuda_dynamic_symbols_t* syms,
+    int default_device_index, iree_allocator_t host_allocator,
+    CUdevice* out_device) {
+  iree_hal_device_info_t* out_device_infos;
+  iree_host_size_t device_count;
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0, iree_hal_cuda_driver_query_available_devices(
+              base_driver, host_allocator, &out_device_infos, &device_count));
   iree_status_t status = iree_ok_status();
   if (device_count == 0 || default_device_index >= device_count) {
     status = iree_make_status(IREE_STATUS_NOT_FOUND,
-                              "default device %d not found (of %d enumerated)",
+                              "default device %d not found (of %ld enumerated)",
                               default_device_index, device_count);
   } else {
```

Comment:
can't add this as a suggestion: 
![image](https://user-images.githubusercontent.com/4010439/135648867-d8a40a3c-0f0c-4120-959d-03570d68d02f.png)

but can you edit this section to this?

```c
  IREE_RETURN_IF_ERROR(iree_hal_cuda_driver_query_available_devices(
      base_driver, host_allocator, &out_device_infos, &device_count));
  iree_status_t status = iree_ok_status();
  if (device_count == 0) {
    status = iree_make_status(IREE_STATUS_UNAVAILABLE,
                              "no compatible CUDA devices were found");
  } else if (default_device_index >= device_count) {
    status = iree_make_status(IREE_STATUS_NOT_FOUND,
                              "default device %d not found (of %ld enumerated)",
                              default_device_index, device_count);
```

* z0 is not defined here, so use the other status macro (this doesn't build with tracing enabled)
* use `IREE_STATUS_UNAVAILABLE` for no devices, `IREE_STATUS_NOT_FOUND` is fine for the other condition

the CTS code skips tests where drivers/devices are unavailable and reports test failures for other statuses (we could adjust that, but it made sense to me at the time)
https://github.com/google/iree/blob/06eb4740c665c442f68c314c03586b595d370827/iree/hal/cts/cts_test_base.h#L47-L55

Reply:
Thanks, I changed the code to match your suggestion.